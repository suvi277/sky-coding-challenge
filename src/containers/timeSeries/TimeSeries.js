import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchTimeSeries } from 'actions';
import Loader from 'components/loader/Loader';
import Chart from 'components/chart/Chart';
import Aux from 'components/hoc/Aux';
import moment from 'moment';

class TimeSeries extends Component {
	componentDidMount() {
		this.props.fetchTimeSeries();
		this.prepareChartData();
	}

	prepareChartData() {
		const { data } = this.props.timeSeries;
		let day;

		let labels = '',
			lineData = '';

		if (data.length) {
			lineData = data.map((time) => time.value);
			labels = data.map((time) => {
				const date = moment(new Date(time.timestamp));
				day = date.format('LL');
				return date.format('mm:ss a');
			});
		}

		return { labels, lineData, day };
	}

	render() {
		const { isLoaded } = this.props.timeSeries;
		return (
			<Aux>
				{!isLoaded && <Loader />}
				{isLoaded && <Chart {...this.prepareChartData()} />}
			</Aux>
		);
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		fetchTimeSeries: () => dispatch(fetchTimeSeries())
	};
};

export default connect(({ timeSeries }) => ({ timeSeries }), mapDispatchToProps)(TimeSeries);
