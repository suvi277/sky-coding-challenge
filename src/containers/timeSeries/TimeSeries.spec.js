import React from 'react';
import { TimeSeries } from './TimeSeries';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import moment from 'moment';

describe('TimeSeries Container', () => {
	let wrapper;

	describe('when renders', () => {
		let props, tree;

		beforeEach(() => {
			props = {
				timeSeries: {
					isLoaded: false,
					data: []
				},
				fetchTimeSeries: jest.fn()
			};
			tree = renderer.create(<TimeSeries {...props} />);
		});

		it('renders as expected', () => {
			expect(tree.toJSON()).toMatchSnapshot();
		});

		it('should call fetchTimeSeries', () => {
			expect(props.fetchTimeSeries).toBeCalled();
		});
	});

	it('render Loader component when time series data is not loaded', () => {
		const props = {
			timeSeries: {
				isLoaded: false,
				data: []
			},
			fetchTimeSeries: jest.fn()
		};
		wrapper = shallow(<TimeSeries {...props} />);
		const Loader = wrapper.find('Loader');
		expect(Loader.length).toBe(1);
	});

	it('render Chart component when time series data is loaded', () => {
		const mockTimeData = [
			{
				timestamp: 1458000000,
				value: 63.422235
			}
		];

		const props = {
			timeSeries: {
				isLoaded: true,
				data: mockTimeData
			},
			fetchTimeSeries: jest.fn()
		};

		wrapper = shallow(<TimeSeries {...props} />);

		const Chart = wrapper.find('Chart');

		const date = moment(new Date(mockTimeData[0].timestamp));
		const day = date.format('LL');
		const labels = [ date.format('mm:ss a') ],
			lineData = [ mockTimeData[0].value ];

		expect(Chart.length).toBe(1);
		expect(Chart.props()).toStrictEqual({ labels, lineData, day });
	});
});
