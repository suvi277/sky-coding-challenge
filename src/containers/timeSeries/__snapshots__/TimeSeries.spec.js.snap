// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSeries Container when renders renders as expected 1`] = `
<div
  className="loader"
>
  <div
    className="spinner-border text-warning"
  />
</div>
`;
